### PR TITLE
Allow access to structured object representation of patch data

### PR DESCRIPTION
**Pull request by bittrance**
This PR does two things: 1) gives control over the context size of patches, and 2) adds a method that allows retrieving a patch in the shape of an object with a list of hunks. E.g.:
```javascript

diff.structuredPatch(
  'oldfile', 'newfile',
  'line2\nline3\nline4\n', 'line2\nline3\nline5',
  'header1', 'header2'
)

{
  oldFileName: 'oldfile', newFileName: 'newfile',
  oldHeader: 'header1', newHeader: 'header2',
  hunks: [{
    oldStart: 1, oldLines: 3, newStart: 1, newLines: 3,
    lines: [' line2', ' line3', '-line4', '+line5', '\\ No newline at end of file'],
  }]
}
```
If it looks acceptable, I'll patch documentation et cetera.

### CODE REVIEW

File: diff.js [hunk 1]
```diff
@@ -398,39 +398,22 @@
...
-        // Figure out if this is the last line for the given file and missing NL
-        if (!(/\n$/.test(current.value)) && (isLast || isLastOfType)) {
-          curRange.push('\\ No newline at end of file');
-        }
-      }
-
+      var hunks = [];
       var oldRangeStart = 0, newRangeStart = 0, curRange = [],
-          oldLine = 1, newLine = 1;
+        oldLine = 1, newLine = 1;
```
Comment (kpdecker) [inline diff.js]: This formatting was intentional. Style is to chain vars with `,` and to double indent when newlines are used to line up the var names.

File: diff.js [hunk 2]
```diff
@@ -454,7 +437,6 @@
           curRange.push.apply(curRange, map(lines, function(entry) {
             return (current.added ? '+' : '-') + entry;
           }));
-          eofNL(curRange, i, current);
```
Comment (kpdecker) [inline diff.js]: Why was this removed?

File: diff.js [hunk 3]
```diff
@@ -466,21 +448,34 @@
...
-                  '@@ -' + oldRangeStart + ',' + (oldLine - oldRangeStart + contextSize)
-                  + ' +' + newRangeStart + ',' + (newLine - newRangeStart + contextSize)
-                  + ' @@');
-              ret.push.apply(ret, curRange);
-              ret.push.apply(ret, contextLines(lines.slice(0, contextSize)));
-              if (lines.length <= 4) {
-                eofNL(ret, i, current);
+              var contextSize = Math.min(lines.length, options.context);
+              var hunklines = [];
+              hunklines.push.apply(hunklines, curRange);
```
Comment (kpdecker) [inline diff.js]: Just do a `curRange.slice()` or even utilize `curRange` if not used elsewhere here.

File: diff.js [hunk 4]
```diff
@@ -466,21 +448,34 @@
...
+              hunklines.push.apply(hunklines, contextLines(lines.slice(0, contextSize)));
+
+              var hunk = { 
+                oldStart: oldRangeStart, 
+                oldLines: (oldLine - oldRangeStart + contextSize), 
+                newStart: newRangeStart, 
+                newLines: (newLine - newRangeStart + contextSize),
+                lines: hunklines
+              }
+              if(i >= diff.length - 2 && lines.length <= options.context) {
```
Comment (kpdecker) [inline diff.js]: Style is to have spaces after `if`, before `(`.

File: diff.js [hunk 5]
```diff
@@ -491,12 +486,40 @@
...
+
+      var ret = [];
+      if (oldFileName == newFileName) {
+        ret.push('Index: ' + oldFileName);
+      }
+      ret.push('===================================================================');
+      ret.push('--- ' + diff.oldFileName + (typeof diff.oldHeader === 'undefined' ? '' : '\t' + diff.oldHeader));
+      ret.push('+++ ' + diff.newFileName + (typeof diff.newHeader === 'undefined' ? '' : '\t' + diff.newHeader));
+
+      for(var i = 0; i < diff.hunks.length; i++) {
```
Comment (kpdecker) [inline diff.js]: Space after `for`

File: diff.js [hunk 6]
```diff
@@ -466,21 +448,34 @@
...
+                lines: hunklines
+              }
+              if(i >= diff.length - 2 && lines.length <= options.context) {
+                // EOF is inside this hunk
+                var oldEOFNewline = /\n$/.test(oldStr);
+                var newEOFNewline = /\n$/.test(newStr);
+                if(lines.length == 0 && !oldEOFNewline) {
+                  hunklines.splice(hunk.oldLines, 0, '\\ No newline at end of file')
+                } else if (!oldEOFNewline || !newEOFNewline) {
+                  hunklines.push('\\ No newline at end of file')
```
Comment (kpdecker) [inline diff.js]: What is the difference between these two cases here?

**Comment by kpdecker** [comment]
This generally looks good outside of a few style concerns and a few cases where I'm not sure what the exact behavior is. Once we get those fixed up, I'm glad to merge.

File: diff.js (hunk 2 above)
Comment (bittrance) [inline diff.js]: I had trouble adapting its logic to the hunk'd format. Since trying to add no-nl each time we have an add/delete entry seemed inefficient, I opted for a rewrite.

File: diff.js (hunk 6 above)
Comment (bittrance) [inline diff.js]: lines.length == 0 && !oldEOFNewline is the special case where no-nl ends up in the middle of hunk. Will add comment to this effect.

**Comment by bittrance** [comment]
Whitespace adjustments done.
Hunklines var removed in favour of curRange.

**Comment by bittrance** [comment]
Documentation added.

**Comment by kpdecker** [comment]
Thanks!

**Comment by kpdecker** [comment]
Released in 2.0.0
